### ISSUE TITLE
Removing the application's data/cache with UIAutomator

### ISSUE THREAD

**Issue by Domuska** [googlesamples/android-testing#98]
Hello!

I'm trying to write a set of tests with UIAutomator, but I ran into a problem while trying to clear the application's data so I could start from a fresh situation at the start of every test.

As mentioned in this [issue](https://github.com/googlesamples/android-testing/issues/54), this was at the time not a feature, is the case still so? If not, could I be pointed to a right direction on how to do this.

What I tried to do was to use this code snippet:

```
   InstrumentationRegistry.getInstrumentation().getUiAutomation()
      .executeShellCommand("pm clear <package_name>")
   .close();
```

but sadly when running multiple tests from Android Studio this results in the first test being ran and then to an error 'Instrumentation run failed due to 'Process crashed.''

I'm not really sure if I'm at all in the right path where I am going with this, or if UIAutomator is even supposed to be used in this manner where a test is ran, a teardown is called and another test is ran on the same application. I know I could use Espresso or Appium or whatnot, but I am just wondering if this kind of thing should be doable with UIAutomator.

**Comment by bootstraponline** [comment]
pm clear terminates the process. The correct way to reset is to define a method on the app and invoke it.

**Comment by Domuska** [comment]
How do you mean define a method on the app? Just have a, let's say a button, that is only in the app's debug version and clears all the app's internal data?

Or is there a way in UIAutomator to directly call methods of the application that is running?

**Comment by bootstraponline** [comment]
If you're using uiautomator v2, then you should be able to call app methods via the instrumentation context. Otherwise, you'll need a "helper" APK which does the clearing.

http://stackoverflow.com/questions/9523564/android-unit-test-clearing-prefs-before-testing-activity/9523741#9523741

**Comment by Domuska** [comment]
Oh yes, I see now. I feel silly not realizing this. I was well able to use my database handler to take care of cleaning up after itself. 

This did raise some questions on how the UIAutomator actually works, but I don't think this would be the right forum to discuss about it.

Thank you a lot for helping with this!

**Comment by afrad** [comment]
fixed in Branch #68
